### PR TITLE
Energy Connectors

### PR DESCRIPTION
**Pull request by LinoxGH**
## Description

Added Energy Connectors!!!

## Changes

Some code refactoring here and there.
Added Energy Connectors with a new research.

## Related Issues

None but this was an approved suggestion.

## Checklist

- [X] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [X] I followed the existing code standards and didn't mess up the formatting.
- [X] I did my best to add documentation to any public classes or methods I added.
- [X] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Review by Sfiguz7 — DISMISSED** [review]
We may want to change the texture to something else before merging but I'm not requesting the change, leaving it up to @TheBusyBiscuit if we want to maybe change the center dots to another color or something of the sort

**Comment by LinoxGH** [comment]
> We may want to change the texture to something else before merging but I'm not requesting the change, leaving it up to @TheBusyBiscuit if we want to maybe change the center dots to another color or something of the sort

I have no idea how to make that so I didn't touch it.

**Comment by LinoxGH** [comment]
Hacktoberfest-accepted label please.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java:291]: Why exactly is this set to null afterwards?

Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java:291]: > Why exactly is this set to null afterwards?

So that we don't bother java gc with this.

**Review by svr333 — DISMISSED** [review]
Ill approve, but Sfiguz said to wait for busys decision, so you'll have to wait for that

**Review by svr333 — DISMISSED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java:754]: This says "Cargo Connector" in the lore...

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java:19]: This line can be removed, it really doesn't have much use here

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java:27]: This variable assignment is completely unnecessary.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java:40]: Still, unnecessary. If you really wanna validate the input, add the plugin as a method parameter and do `Validate.notNull()` but not this.

I think using `SlimefunPlugin.instance()` like before was just fine too.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java:754]: Yeah this will definitely need a proper texture first.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/EnergyNetComponent.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/EnergyNetComponent.java:69]: eehhh, not exactly great but I get why it needs to be here for this.
Can you at least refactor it?

```suggestion

        // Emergency fallback, this cannot hold a charge, so we'll just return zero
        if (!isChargeable()) {
            return 0;
        }
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java
Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/ResearchSetup.java]: I just thought that getting using a field would be more light-weight than running the method over 250 times.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java
Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java]: I'd love to only if I knew how to do it... Well at least without stealing someone's texture and still keeping my account's skin.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java
Comment (Sfiguz7) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java:753]: Only thing missing is changing the texture here, I think you created one and posted it somewhere?
Just figured I'd review the changes and bump this since you requested a review a couple of days ago

**Review by Sfiguz7 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java
Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunItems.java:753]: > Only thing missing is changing the texture here, I think you created one and posted it somewhere?
> Just figured I'd review the changes and bump this since you requested a review a couple of days ago

Yeah the texture is ready. I'll add it to the code and to mojand database soon.

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: Spot the difference...

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java
Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: Blame @J3fftw1 his tutorial gave this result.

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: Dont tag me also you could have used the shorter hash

Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: You need the hex texture id though, not the base64 json properties.
Simply decode this base64 mess, copy the texture url and snip off the domain. Then you get the hex id.

For the future I would recommend you to just use a tool like https://minecraft-heads.com/custom-heads/heads-generator

Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: So I need a base64 of https://textures.minecraft.net/texture/1085e098756b995b00241644089c55a8f9acde35b9a37785d5e057a923613b ?

Comment (TheBusyBiscuit) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: No you just need the hex, the last part of this url...
10853...613b

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: 1085e098756b995b00241644089c55a8f9acde35b9a37785d5e057a923613b

Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: @J3fftw1 btw this is another reason to not use that tutorial.

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/HeadTexture.java:68]: Okay idc if you cant use it and stop tagging me

**Review by TheBusyBiscuit — APPROVED** [review]
Still not the greatest fan of it but the code looks good and people approved this, so... LGTM

Only idea on how to improve it would be to include the connectors in the capacitor recipes since they basically function the same but can also hold energy.

**Comment by LinoxGH** [comment]
> Still not the greatest fan of it but the code looks good and people approved this, so... LGTM
> 
> Only idea on how to improve it would be to include the connectors in the capacitor recipes since they basically function the same but can also hold energy.

Nice idea actually. Gonna do that.

**Review by svr333 — DISMISSED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java:1509]: ugh, I didn't even really see the recipe. That's a lot of CARBON...

How about this?
```
CWC
WRW
CWC
````
C = Carbon
W = Copper WIRE
R = Redstone block

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java
Comment (LinoxGH) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java:1509]: Well... Insulation lol

**Review by TheBusyBiscuit — APPROVED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java
Comment (svr333) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/setup/SlimefunItemSetup.java]: how about 4 carbon (empty corners)?
And honestly I dont think many people would make this cuz you can just put another capacitator there - makes it less clustered overall

**Review by svr333 — APPROVED** [review]

**Pull request merged**
